libp2p/Host.cpp

```diff
@@ -367,22 +372,233 @@ void Host::runAcceptor()
 	}
 }
 
-void Host::doHandshake(bi::tcp::socket* _socket, NodeId _nodeId)
+void PeerHandshake::transition(boost::system::error_code _ech)
 {
-	try {
-		clog(NetConnect) << "Accepting connection for " << _socket->remote_endpoint();
-	} catch (...){}
-
-	shared_ptr<Peer> p;
-	if (_nodeId)
-		p = m_peers[_nodeId];
+	if (_ech)
+	{
+		boost::system::error_code ec;
+		socket->shutdown(boost::asio::ip::tcp::socket::shutdown_both, ec);
+		if (socket->is_open())
+			socket->close();
+		return;
+	}
 	
-	if (!p)
-		p.reset(new Peer());
-	p->endpoint.tcp.address(_socket->remote_endpoint().address());
+	auto self(shared_from_this());
+	if (nextState == New)
+	{
+		nextState = AckAuth;
+		
+		clog(NetConnect) << "Authenticating connection for " << socket->remote_endpoint();
+		
+		if (originated)
+		{
+			clog(NetConnect) << "devp2p.connect.egress sending auth";
+			// egress: tx auth
+			asserts((bool)remote);
```

Comment:
If I understand this correctly, `remote` is a `h512` then you don't really want a c-style cast to `bool` here. If you must, try double negation.